pallets/block-rewards-hybrid/src/lib.rs

```diff
@@ -0,0 +1,384 @@
+// This file is part of Astar.
+
+// Copyright (C) 2019-2023 Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+//! # Block Reward Distribution Pallet
+//!
+//! - [`Config`]
+//!
+//! ## Overview
+//!
+//! Pallet that implements block reward issuance and distribution mechanics.
+//!
+//! After issuing a block reward, pallet will calculate how to distribute the reward
+//! based on configurable parameters and chain state.
+//!
+//! Major on-chain factors which can influence reward distribution are total issuance and total value locked by dapps staking.
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Function
+//!
+//! - `set_configuration` - used to change reward distribution configuration parameters
+//!
+//! ### Other
+//!
+//! - `on_timestamp_set` - This pallet implements the `OnTimestampSet` trait to handle block production.
+//!                        Note: We assume that it's impossible to set timestamp two times in a block.
+//!
+//! ## Usage
+//!
+//! 1. Pallet should be set as a handler of `OnTimestampSet`.
+//! 2. `DappsStakingTvlProvider` handler should be defined as an impl of `TvlProvider` trait. For example:
+//! ```nocompile
+//! pub struct TvlProvider();
+//! impl Get<Balance> for TvlProvider {
+//!     fn tvl() -> Balance {
+//!         DappsStaking::total_locked_value()
+//!     }
+//! }
+//! ```
+//! 3. `BeneficiaryPayout` handler should be defined as an impl of `BeneficiaryPayout` trait. For example:
+//! ```nocompile
+//! pub struct BeneficiaryPayout();
+//! impl BeneficiaryPayout<NegativeImbalanceOf<T>> for BeneficiaryPayout {
+//!
+//!     fn treasury(reward: NegativeImbalanceOf<T>) {
+//!         Balances::resolve_creating(&TREASURY_POT.into_account(), reward);
+//!     }
+//!
+//!     fn collators(reward: NegativeImbalanceOf<T>) {
+//!         Balances::resolve_creating(&COLLATOR_POT.into_account(), reward);
+//!      }
+//!
+//!     fn dapps_staking(stakers: NegativeImbalanceOf<T>, dapps: NegativeImbalanceOf<T>) {
+//!         DappsStaking::rewards(stakers, dapps);
+//!     }
+//! }
+//! ```
+//! 4. Set `RewardAmount` to desired block reward value in native currency.
```

Comment:
This should be updated with new `MaxBlockRewardAmount` config and cover the mechanism.